### PR TITLE
Feature/add generic controller

### PR DESCRIPTION
**Pull request by Blackolf931**

### CODE REVIEW

File: TSP/BLL/Services/EmployeeService.cs [hunk 1]
```diff
@@ -20,12 +20,12 @@ public EmployeeService(IRepositoryBase<EmployeeEntity> repository, IMapper mappe
...
         }
 
-        public override async Task<Employee?> GetByIdAsync(int id)
+        public override async Task<Employee> GetByIdAsync(int id)
         {
             var entityObject = await _repository.FindByIdAsync(id);
             if (entityObject is null)
             {
-                return null;
+                return _mapper.Map<Employee>(null);
```
Comment (nantonov) [inline TSP/BLL/Services/EmployeeService.cs:28]: Why?

File: TSP/BLL/Services/GenericServiceSetId.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿using BLL.Interfaces;
+
+namespace BLL.Services
+{
+    public class GenericServiceSetId<T> : ISetId<T>
+    {
+        public virtual T SetId(int id, T model)
```
Comment (nantonov) [inline TSP/BLL/Services/GenericServiceSetId.cs:7]: [WTF part] What does this method do?

File: TSP/TSP.API/Validators/EmployeeViewModelValidator.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using FluentValidation;
+using TSP.API.ViewModels;
+
+namespace TSP.API.Validators
+{
+    public class EmployeeViewModelValidator : AbstractValidator<EmployeeViewModel>
```
Comment (nantonov) [inline TSP/TSP.API/Validators/EmployeeViewModelValidator.cs:6]: For what you validate this vm?

File: TSP/BLL/Services/SetOfficeId.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using BLL.Interfaces;
+using BLL.Models;
+
+namespace BLL.Services
+{
+    public class SetOfficeId : GenericServiceSetId<Office>, IOfficeSetId
```
Comment (nantonov) [inline TSP/BLL/Services/SetOfficeId.cs:6]: Strange name

File: TSP/BLL/Services/SetEmployeeId.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using BLL.Interfaces;
+using BLL.Models;
+namespace BLL.Services
+{
+    public class SetEmployeeId : GenericServiceSetId<Employee>, IEmployeeSetId
```
Comment (nantonov) [inline TSP/BLL/Services/SetEmployeeId.cs:5]: Strange name

File: TSP/BLL/Services/GenericService.cs [hunk 6]
```diff
@@ -35,26 +32,23 @@ public async Task<bool> DeleteByIdAsync(int id)
...
         }
-
         public virtual async Task<T> GetByIdAsync(int id)
         {
             var entityObject = await _repository.FindByIdAsync(id);
             var mappedEmployee = _mapper.Map<T>(entityObject);
             return mappedEmployee;
         }
-        public async Task<T> UpdateAsync(T entity)
+        public virtual async Task<T> UpdateAsync(T entity)
```
Comment (nantonov) [inline TSP/BLL/Services/GenericService.cs:46]: Why it virtual?

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/Services/GenericService.cs (hunk 6 above)
Comment (Blackolf931) [reply TSP/BLL/Services/GenericService.cs]: fix

File: TSP/BLL/Services/SetEmployeeId.cs (hunk 5 above)
Comment (Blackolf931) [reply TSP/BLL/Services/SetEmployeeId.cs]: deleted

File: TSP/BLL/Services/SetOfficeId.cs (hunk 4 above)
Comment (Blackolf931) [reply TSP/BLL/Services/SetOfficeId.cs]: deleted

File: TSP/TSP.API/Validators/EmployeeViewModelValidator.cs (hunk 3 above)
Comment (Blackolf931) [reply TSP/TSP.API/Validators/EmployeeViewModelValidator.cs]: fix

File: TSP/BLL/Services/GenericServiceSetId.cs (hunk 2 above)
Comment (Blackolf931) [reply TSP/BLL/Services/GenericServiceSetId.cs]: deleted

File: TSP/BLL/Services/EmployeeService.cs (hunk 1 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmployeeService.cs]: fix

File: TSP/BLL/Services/IEmployeeRepository.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+﻿namespace BLL.Services
+{
+    internal interface IEmployeeRepository<T>
```
Comment (nantonov) [inline TSP/BLL/Services/IEmployeeRepository.cs:3]: empty employee generic repo in BLL?

File: TSP/DAL/DI/DependencyDalRegistrar.cs [hunk 8]
```diff
@@ -11,7 +11,8 @@ public static class DependencyDalRegistrar
     {
         public static void RegistarDalComponents(this IServiceCollection services, IConfiguration configuration)
         {
-            services.AddScoped(typeof(IRepositoryBase<>), typeof(RepositoryBase<>));
+            services.AddScoped<IEmployeeRepository, EmployeeRepository>();
+            services.AddScoped<IOfficeRepository, OfficeRepository>(); ;
```
Comment (nantonov) [inline TSP/DAL/DI/DependencyDalRegistrar.cs:15]: ;

File: TSP/TSP.API/Validators/ViewModelAddOfficeValidator.cs [hunk 9]
```diff
@@ -7,9 +7,9 @@ public class ViewModelAddOfficeValidator : AbstractValidator<OfficeAddViewModel>
     {
         public ViewModelAddOfficeValidator()
         {
-            RuleFor(x => x.Name).NotEmpty().Matches("^[A-Z][a-zA-Z]*$").WithMessage("Name can't be empty").MaximumLength(60).MinimumLength(3);
-            RuleFor(x => x.Address).NotEmpty().Matches("^[A - Za - z0 - 9'\\.\\-\\s\\,]").MaximumLength(60).MinimumLength(3);
-            RuleFor(x => x.Country).NotEmpty().Matches("^[A-Z][a-zA-Z]*$").MaximumLength(20);
+         //   RuleFor(x => x.Name).NotEmpty().Matches("^[A-Z][a-zA-Z]*$").WithMessage("Name can't be empty").MaximumLength(60).MinimumLength(3);
```
Comment (nantonov) [inline TSP/TSP.API/Validators/ViewModelAddOfficeValidator.cs:10]: Why?

File: TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs [hunk 10]
```diff
@@ -1,8 +1,11 @@
-﻿namespace TSP.API.ViewModels
+﻿using TSP.API.Interfaces;
+
+namespace TSP.API.ViewModels
 {
 #nullable disable
-    public class EmployeeAddViewModel
+    public class EmployeeAddViewModel : IHasIdBase
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs:6]: You cannot insert Id

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs [hunk 11]
```diff
@@ -1,11 +1,14 @@
-﻿namespace TSP.API.ViewModels
+﻿using TSP.API.Interfaces;
+
+namespace TSP.API.ViewModels
 {
 #nullable disable
-    public class OfficeAddViewModel
+    public class OfficeAddViewModel : IHasIdBase
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/OfficeAddViewModel.cs:6]: You cannot insert Id

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/Services/IEmployeeRepository.cs (hunk 7 above)
Comment (Blackolf931) [reply TSP/BLL/Services/IEmployeeRepository.cs]: deleted

File: TSP/DAL/DI/DependencyDalRegistrar.cs (hunk 8 above)
Comment (Blackolf931) [reply TSP/DAL/DI/DependencyDalRegistrar.cs]: fix

File: TSP/TSP.API/Validators/ViewModelAddOfficeValidator.cs (hunk 9 above)
Comment (Blackolf931) [reply TSP/TSP.API/Validators/ViewModelAddOfficeValidator.cs]: fix

File: TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs (hunk 10 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs]: fix

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs (hunk 11 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/OfficeAddViewModel.cs]: fix

File: TSP/TSP.API/Controllers/GenericController.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            await _validatorAddViewModel.ValidateAndThrowAsync(addViewModel);
+            var mapped = _mapper.Map<T>(addViewModel);
+            return Ok(await _service.AddAsync(mapped));
+        }
+        [HttpPut("Update/{id}")]
+        public virtual async Task<ActionResult<T>> UpdateAsync(int id, [FromBody] TUpdateViewModel viewModel)
+        {
+            await _validatorUpdateViewModel.ValidateAndThrowAsync(viewModel);
+            viewModel.Id = id;
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/GenericController.cs:55]: You can set Id to Model, not ViewModel

File: TSP/TSP.API/Validators/ViewModelAddEmployeeValidator.cs [hunk 13]
```diff
@@ -3,9 +3,9 @@
 
 namespace TSP.API.Validators
 {
-    public class EmployeeAddViewModelValidator : AbstractValidator<EmployeeAddViewModel>
+    public class ViewModelAddEmployeeValidator : AbstractValidator<EmployeeAddViewModel>
```
Comment (nantonov) [inline TSP/TSP.API/Validators/ViewModelAddEmployeeValidator.cs:6]: What happened with naming?

File: TSP/TSP.API/Validators/ViewModelUpdateEmployeeValidator.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
+﻿using FluentValidation;
+using TSP.API.ViewModels;
+
+namespace TSP.API.Validators
+{
+    public class ViewModelUpdateEmployeeValidator : AbstractValidator<EmployeeUpdateViewModel>
```
Comment (nantonov) [inline TSP/TSP.API/Validators/ViewModelUpdateEmployeeValidator.cs:6]: Naming

File: TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs [hunk 15]
```diff
@@ -3,6 +3,7 @@
 #nullable disable
     public class EmployeeAddViewModel
     {
+        public int Id { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs:6]: We cannot add entity with id

File: TSP/TSP.API/ViewModels/EmployeeUpdateViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
+﻿using TSP.API.Interfaces;
+
+namespace TSP.API.ViewModels
+{
+#nullable disable
+    public class EmployeeUpdateViewModel : IHasIdBase
+    {
+        public int Id { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeUpdateViewModel.cs:8]: User send the Id in the query

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs [hunk 17]
```diff
@@ -6,6 +6,7 @@ public class OfficeAddViewModel
         public string Name { get; set; }
         public string Address { get; set; }
         public string Country { get; set; }
+        public int Id { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/OfficeAddViewModel.cs:9]: the same

File: TSP/TSP.API/ViewModels/OfficeUpdateViewModel.cs [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace TSP.API.ViewModels
+{
+#nullable disable
+    public class OfficeUpdateViewModel : IHasIdBase
+    {
+        public string Name { get; set; }
+        public string Address { get; set; }
+        public string Country { get; set; }
+        public int Id { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/OfficeUpdateViewModel.cs:11]: will be send through query

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/TSP.API/Controllers/GenericController.cs (hunk 12 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/GenericController.cs]: fix

File: TSP/TSP.API/Validators/ViewModelAddEmployeeValidator.cs (hunk 13 above)
Comment (Blackolf931) [reply TSP/TSP.API/Validators/ViewModelAddEmployeeValidator.cs]: fix

File: TSP/TSP.API/Validators/ViewModelUpdateEmployeeValidator.cs (hunk 14 above)
Comment (Blackolf931) [reply TSP/TSP.API/Validators/ViewModelUpdateEmployeeValidator.cs]: fix

File: TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs (hunk 15 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/EmployeeAddViewModel.cs]: fix

File: TSP/TSP.API/ViewModels/EmployeeUpdateViewModel.cs (hunk 16 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/EmployeeUpdateViewModel.cs]: fix

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs (hunk 17 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/OfficeAddViewModel.cs]: fix

File: TSP/TSP.API/ViewModels/OfficeUpdateViewModel.cs (hunk 18 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/OfficeUpdateViewModel.cs]: fix

**Review by nantonov — APPROVED** [review]

**Pull request merged**